gradle-baseline-java/src/main/groovy/com/palantir/baseline/plugins/BaselineModuleJvmArgs.java

```diff
@@ -0,0 +1,214 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.baseline.plugins;
+
+import com.google.common.base.Splitter;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.palantir.baseline.extensions.BaselineExportsExtension;
+import java.io.IOException;
+import java.util.Set;
+import java.util.jar.JarFile;
+import java.util.stream.Stream;
+import org.gradle.api.Action;
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.gradle.api.Task;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.java.archives.Manifest;
+import org.gradle.api.tasks.JavaExec;
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.SourceSetContainer;
+import org.gradle.api.tasks.compile.JavaCompile;
+import org.gradle.api.tasks.testing.Test;
+import org.gradle.jvm.tasks.Jar;
+import org.gradle.process.CommandLineArgumentProvider;
+
+/**
+ * This plugin reuses the {@code Add-Exports} manifest entry defined in
+ * <a href="https://openjdk.java.net/jeps/261">JEP-261</a> to propagate and collect required exports
+ * from transitive dependencies, and applies them to compilation (for annotation processors) and
+ * execution (tests, javaExec, etc) for runtime dependencies.
+ */
+public final class BaselineModuleJvmArgs implements Plugin<Project> {
+
+    private static final String EXTENSION_NAME = "moduleJvmArgs";
+    private static final String ADD_EXPORTS_ATTRIBUTE = "Add-Exports";
+
+    private static final Splitter EXPORT_SPLITTER =
+            Splitter.on(' ').trimResults().omitEmptyStrings();
+
+    @Override
+    public void apply(Project project) {
+        project.getPluginManager().withPlugin("java", unused -> {
+            BaselineExportsExtension extension =
+                    project.getExtensions().create(EXTENSION_NAME, BaselineExportsExtension.class, project);
+
+            // javac isn't provided `--add-exports` args for the time being due to
+            // https://github.com/gradle/gradle/issues/18824
+            if (project.hasProperty("add.exports.to.javac")) {
+                project.getExtensions().getByType(SourceSetContainer.class).configureEach(sourceSet -> {
+                    JavaCompile javaCompile = project.getTasks()
+                            .named(sourceSet.getCompileJavaTaskName(), JavaCompile.class)
+                            .get();
+                    javaCompile
+                            .getOptions()
+                            .getCompilerArgumentProviders()
+                            // Use an anonymous class because tasks with lambda inputs cannot be cached
+                            .add(new CommandLineArgumentProvider() {
```

Comment:
As a nitty suggestion, all three `CommandLineArgumentProvider` are quite similar except for the `arguments = ...` line. To make this more concise, we might we can define this as a single static class that gets a `Provider<List<String>> arguments` injected in its constructor?

Reply:
Thoughts on punting this? It would mean plumbing around logging args to differentiate the types of tasks as well

Reply:
Yeah, no strong opinion on that one!